packages/xod-project/src/patch.js

```diff
@@ -1366,3 +1380,140 @@ export const validateAbstractPatch = def(
+      )
+    ) {
+      return fail('SPECIALIZATION_PATCH_MUST_HAVE_SAME_ARITY_LEVEL', {});
+    }
+
+    const checkedPatchDoesHaveGenericPins = R.compose(
+      R.any(Pin.isGenericPin),
+      listPins
+    )(specializationPatch);
+
+    if (checkedPatchDoesHaveGenericPins) {
+      return fail('SPECIALIZATION_PATCH_CANT_HAVE_GENERIC_PINS', {});
+    }
+
+    const [
+      numberOfPinsInAbstractPatchByDirection,
+      numberOfPinsInSpecializationPatchByDirection,
+    ] = R.map(
+      R.compose(R.map(R.length), R.groupBy(Pin.getPinDirection), listPins)
+    )([abstractPatch, specializationPatch]);
+
+    // check that patch has a proper number of inputs and outputs
+    if (
+      numberOfPinsInAbstractPatchByDirection[CONST.PIN_DIRECTION.INPUT] !==
+      numberOfPinsInSpecializationPatchByDirection[CONST.PIN_DIRECTION.INPUT]
+    ) {
+      return fail('SPECIALIZATION_PATCH_MUST_HAVE_N_INPUTS', {
+        desiredInputsNumber:
+          numberOfPinsInAbstractPatchByDirection[CONST.PIN_DIRECTION.INPUT],
+      });
+    }
+
+    if (
+      numberOfPinsInAbstractPatchByDirection[CONST.PIN_DIRECTION.OUTPUT] !==
+      numberOfPinsInSpecializationPatchByDirection[CONST.PIN_DIRECTION.OUTPUT]
+    ) {
+      return fail('SPECIALIZATION_PATCH_MUST_HAVE_N_OUTPUTS', {
+        desiredOutputsNumber:
+          numberOfPinsInAbstractPatchByDirection[CONST.PIN_DIRECTION.OUTPUT],
+      });
+    }
+
+    const [genericPinPairs, staticPinPairs] = R.compose(
+      R.partition(R.pipe(R.nth(0), Pin.isGenericPin)),
+      R.apply(R.zip),
+      R.map(
+        R.compose(
+          R.sortWith([
+            R.ascend(Pin.getPinDirection),
+            R.ascend(Pin.getPinOrder),
+          ]),
+          listPins
+        )
+      )
+    )([abstractPatch, specializationPatch]);
+
+    const staticTypesMatch = R.all(
+      R.apply(R.eqBy(Pin.getPinType)),
+      staticPinPairs
+    );
+    // TODO: `find` instead of `any` and explain which one does not match?
+    if (!staticTypesMatch) {
+      return fail('SPECIALIZATION_STATIC_PINS_DO_NOT_MATCH', {});
+    }
+
+    const ambiguousGenerics = R.compose(
+      R.toPairs,
+      R.filter(resolutions => resolutions.length > 1),
+      R.map(R.pipe(R.map(R.nth(1)), R.uniqBy(Pin.getPinType))),
+      R.groupBy(R.pipe(R.nth(0), Pin.getPinType))
+    )(genericPinPairs);
+
+    if (!R.isEmpty(ambiguousGenerics)) {
+      const [genericType, pinsWithDifferentTypes] = ambiguousGenerics[0];
+      const typeNames = R.compose(R.join(', '), R.map(Pin.getPinType))(
+        pinsWithDifferentTypes
+      );
+
+      return fail('SPECIALIZATION_HAS_CONFLICTING_TYPES_FOR_GENERIC', {
+        genericType,
+        typeNames,
+      });
+    }
+
+    const abstractPatchBaseName = R.compose(getBaseName, getPatchPath)(
+      abstractPatch
+    );
+    const expectedSpecializationBaseName = R.compose(
+      commaSeparatedTypes => `${abstractPatchBaseName}(${commaSeparatedTypes})`,
```

Comment:
Looks like some kind of magic over specialization patchPaths here.
How about to extract it into the separate function in the `patchPathUtils.js`?

E.G.
`getSpecializationPatchPath :: PatchPath -> [DataType] -> PatchPath`
(but actually it is not ours `DataType`)
and then do `getBaseName` here.

I think it will make maintaining easier if we want to change something with patch paths in the future.